src/util/CircularBuffer.h

```diff
@@ -0,0 +1,75 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include <vector>
+
+namespace stellar
+{
+
+using std::vector;
+using std::size_t;
+
+template <typename T> class CircularBuffer
+{
+  public:
+    explicit CircularBuffer(const size_t size) : buffer(size), count(0), start(0)
+    {
+    }
+
+    void
+    push(T elem)
+    {
+        size_t ix = postIncrementLastIndex();
+        buffer[ix] = elem;
+    }
+
+    T&
+    pop()
+    {
+        if (empty())
+        {
+            return buffer[0];
+        }
+        size_t start = postIncrementStartIndex();
+        --count;
+        return buffer[start];
+    }
+
+    bool
+    empty() const
+    {
+        return 0 == count;
+    }
+
+  private:
+    vector<T> buffer;
+    size_t count;
+    size_t start;
+
+    size_t
+    postIncrementLastIndex()
```

Comment:
this method looks like its name does not reflect what it is doing:
- compute new last index
- increment count if needed
- change start index if needed

this could be probably simplified by introducing two indexes, begin and end
then adding item would be as simple as:

buffer[end] = newItem;
end = (end + 1) % size;
if (end == begin)
  begin = (begin + 1) % size

and removing:
auto popped = buffer[begin];
begin = (begin + 1) % size

and empty = (begin == end)
and (x + 1) % size could be refactored

If you like this, please change. If not, leave it as it is now.